File: ClosestPointPair/ClosestPointPair.cpp
#include <fstream>
#include <vector>
#include <sstream>
#include <algorithm>
#include <cmath>
#include "ClosestPointPair.h"

using namespace std;

struct ClosestPointPair::Point {
    double x;
    double y;

    bool operator<(Point const &other) const {
        return y < other.y;
    }
};

bool ClosestPointPair::sortX(const Point &a, const Point &b) {
    return a.x < b.x;
}

vector<string> ClosestPointPair::parseInput(const string &filePath) {
    vector<string> rawPoints;
    ifstream file(filePath);
    string input;
    while (getline(file, input, '}')) {
        if (input.length() > 3) {
            if (input.find("{{") != string::npos) {
                input = input.substr(2);
            } else {
                input = input.substr(3);
            }
            rawPoints.push_back(input);
        }
    }
    return rawPoints;
}

void ClosestPointPair::populatePoints(Point* points, int pointsCount, vector<string> rawPoints) {
    for (int i = 0; i < pointsCount; ++i) {
        Point point{};
        string part;
        stringstream ss(rawPoints[i]);
        getline(ss, part, ',');
        point.x = stod(part);
        getline(ss, part, ',');
        point.y = stod(part.substr(1));
        points[i] = point;
    }
}

double ClosestPointPair::splitAndSolve(Point* points, int pointsCount) {
    if (pointsCount < 10) {
        double bruteForceMin = sqrt(
            pow(points[0].x - points[1].x, 2) + pow(points[0].y - points[1].y, 2)
        );
        for (int i = 0; i < pointsCount; i++) {
            for (int j = i + 1; j < pointsCount; j++) {
                double distance = sqrt(
                    pow(points[i].x - points[j].x, 2) +
                    pow(points[i].y - points[j].y, 2)
                );
                if (bruteForceMin > distance) {
                    bruteForceMin = distance;
                }
            }
        }
        return bruteForceMin;
    }
    int middleIndex = pointsCount / 2;
    Point* leftPoints = points;
    Point* rightPoints = points + middleIndex;  // Might need a +1 here
    double newMin = min(splitAndSolve(leftPoints, middleIndex), splitAndSolve(rightPoints, middleIndex));
    int leftIndex = middleIndex;
    double leftMin = points[middleIndex].x - newMin;
    while (points[leftIndex].x >= leftMin && leftIndex > 0) {
        leftIndex--;
    }
    int rightIndex = middleIndex;
    double rightMax = points[middleIndex].x + newMin;
    while (points[rightIndex].x <= rightMax && rightIndex < pointsCount) {
        rightIndex++;
    }
    Point* middlePoints = points + leftIndex;
    int middleSize = rightIndex - leftIndex;
    sort(middlePoints, middlePoints + middleSize);
    for (int i = 0; i < middleSize; i++) {
        for (int j = 1; j <= 2; j++) {
            if (i + j < middleSize) {
                double distance = sqrt(
                    pow(middlePoints[i].x - middlePoints[i + j].x, 2) +
                    pow(middlePoints[i].y - middlePoints[i + j].y, 2)
                );
                if (distance < newMin) {
                    newMin = distance;
                }
            }
        }
    }
    return newMin;
}

double ClosestPointPair::divideAndConquer(const string &filePath) {
    vector<string> rawPoints = parseInput(filePath);
    int pointsCount = int(rawPoints.size());
    auto* points = new Point[pointsCount];
    populatePoints(points, pointsCount, rawPoints);
    sort(points, points + pointsCount, sortX);
    double ans = round(splitAndSolve(points, pointsCount) * 1000) / 1000;
    delete [] points;
    return ans;
}

double ClosestPointPair::bruteForce(const string &filePath) {
    vector<string> rawPoints = parseInput(filePath);
    int pointsCount = int(rawPoints.size());
    auto* points = new Point[pointsCount];
    populatePoints(points, pointsCount, rawPoints);
    double bruteForceMin = sqrt(
        pow(points[0].x - points[1].x, 2) + pow(points[0].y - points[1].y, 2)
    );
    for (int i = 0; i < pointsCount; i++) {
        for (int j = i + 1; j < pointsCount; j++) {
            double distance = sqrt(
                pow(points[i].x - points[j].x, 2) +
                pow(points[i].y - points[j].y, 2)
            );
            if (bruteForceMin > distance) {
                bruteForceMin = distance;
            }
        }
    }
    delete [] points;
    return round(bruteForceMin * 1000) / 1000;
}

File: ClosestPointPair/ClosestPointPair.py
import math


def split_and_solve(points):
    if len(points) < 10:
        brute_min = float('inf')
        for i, p1 in enumerate(points):
            for p2 in points[i + 1:]:
                distance = math.dist(p1, p2)
                if brute_min > distance:
                    brute_min = distance
        return brute_min
    middle_idx = len(points) // 2
    new_min = min(split_and_solve(points[:middle_idx]), split_and_solve(points[middle_idx:]))
    left_idx = middle_idx
    left_min = points[middle_idx][0] - new_min
    while points[left_idx][0] >= left_min and left_idx > 0:
        left_idx -= 1
    right_idx = middle_idx
    right_max = points[middle_idx][0] + new_min
    while right_idx < len(points) and points[right_idx][0] <= right_max:
        right_idx += 1
    middle_points = sorted(points[left_idx:right_idx], key=lambda x: x[1])
    for i, p1 in enumerate(middle_points):
        for j in range(1, 3):
            if i + j < len(middle_points):
                distance = math.dist(p1, middle_points[i + j])
                if distance < new_min:
                    new_min = distance
    return new_min


def divide_and_conquer(file_path):
    points = sorted(
        [[float(i) for i in a.split(', ')] for a in open(file_path).read()[2:-2].split('}, {')],
        key=lambda c: c[0]
    )
    return round(split_and_solve(points), 3)


def brute_force(file_path):
    points = sorted(
        [[float(i) for i in a.split(', ')] for a in open(file_path).read()[2:-2].split('}, {')],
        key=lambda c: c[0]
    )
    new_min = float('inf')
    for i, p1 in enumerate(points):
        for p2 in points[i + 1:]:
            distance = math.dist(p1, p2)
            if new_min > distance:
                new_min = distance
    return new_min
